### PR TITLE
Feature/test serial

### PR DESCRIPTION
**Pull request by gustavosinbandera1**

### CODE REVIEW

File: lib/cmd_nvs/src/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+set(COMPONENT_ADD_INCLUDEDIRS .)
+
+set(COMPONENT_SRCS "cmd_nvs.c")
```
Comment (luisan00) [inline lib/cmd_nvs/src/CMakeLists.txt]: Rename `cmd_nvs.c` to `cmd_nvs.cpp`

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 2]
```diff
@@ -0,0 +1,518 @@
+/* Console example — NVS commands
+
+   This example code is in the Public Domain (or CC0 licensed, at your option.)
+
+   Unless required by applicable law or agreed to in writing, this
+   software is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+   CONDITIONS OF ANY KIND, either express or implied.
+*/
+
+#include <stdio.h>
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: When possible, use `cstdio.h` instead

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+#include <errno.h>
+#include <string.h>
+#include <stdlib.h>
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: When possible, use `cstdlib ` instead

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+   This example code is in the Public Domain (or CC0 licensed, at your option.)
+
+   Unless required by applicable law or agreed to in writing, this
+   software is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+   CONDITIONS OF ANY KIND, either express or implied.
+*/
+
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: When possible, use `cstring` instead

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 5]
```diff
@@ -0,0 +1,518 @@
+/* Console example — NVS commands
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: also add the common header of the project

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+}
+
+static void print_blob(const char *blob, size_t len)
+{
+    for (int i = 0; i < len; i++) {
+        printf("%02x", blob[i]);
+    }
+    printf("\n");
+}
+
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: The documentation header is missing

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+    esp_err_t err = nvs_set_blob(nvs, key, blob, blob_len);
+    free(blob);
+
+    if (err == ESP_OK) {
+        err = nvs_commit(nvs);
+    }
+
+    return err;
+}
+
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: The documentation header is missing

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 8]
```diff
@@ -0,0 +1,518 @@
...
+        err = nvs_commit(nvs);
+        if (err == ESP_OK) {
+            ESP_LOGI(TAG, "Value stored under key '%s'", key);
+        }
+    }
+
+    nvs_close(nvs);
+    return err;
+}
+
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: The documentation header is missing

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 9]
```diff
@@ -0,0 +1,518 @@
...
+    struct arg_str *namespace;
+    struct arg_end *end;
+} erase_all_args;
+
+static struct {
+    struct arg_str *namespace;
+    struct arg_end *end;
+} namespace_args;
+
+
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: The documentation header is missing

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 10]
```diff
@@ -0,0 +1,518 @@
...
+                print_blob(blob, len);
+            }
+            free(blob);
+        }
+    }
+
+    nvs_close(nvs);
+    return err;
+}
+
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: The documentation header is missing

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 11]
```diff
@@ -0,0 +1,518 @@
...
+            if (err == ESP_OK) {
+                ESP_LOGI(TAG, "Value with key '%s' erased", key);
+            }
+        }
+        nvs_close(nvs);
+    }
+
+    return err;
+}
+
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: The documentation header is missing

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 12]
```diff
@@ -0,0 +1,518 @@
...
+        if (err == ESP_OK) {
+            err = nvs_commit(nvs);
+        }
+    }
+
+    ESP_LOGI(TAG, "Namespace '%s' was %s erased", name, (err == ESP_OK) ? "" : "not");
+    nvs_close(nvs);
+    return ESP_OK;
+}
+
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: The documentation header is missing

File: lib/cmd_system/src/cmd_system.c [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+        .command = "version",
+        .help = "Get version of chip and SDK",
+        .hint = NULL,
+        .func = &get_version,
+    };
+    ESP_ERROR_CHECK( esp_console_cmd_register(&cmd) );
+}
+
+/** 'restart' command restarts the program */
+
```
Comment (luisan00) [inline lib/cmd_system/src/cmd_system.c]: The documentation header is missing

File: lib/cmd_system/src/cmd_system.c [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+    printf("\tfeature:%s%s%s%s%d%s\r\n",
+           info.features & CHIP_FEATURE_WIFI_BGN ? "/802.11bgn" : "",
+           info.features & CHIP_FEATURE_BLE ? "/BLE" : "",
+           info.features & CHIP_FEATURE_BT ? "/BT" : "",
+           info.features & CHIP_FEATURE_EMB_FLASH ? "/Embedded-Flash:" : "/External-Flash:",
+           spi_flash_get_chip_size() / (1024 * 1024), " MB");
+    printf("\trevision number:%d\r\n", info.revision);
+    return 0;
+}
+
```
Comment (luisan00) [inline lib/cmd_system/src/cmd_system.c]: The documentation header is missing

File: lib/cmd_system/src/cmd_system.c [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+    register_version();
+    register_restart();
+    register_deep_sleep();
+    register_light_sleep();
+#if WITH_TASKS_INFO
+    register_tasks();
+#endif
+}
+
+/* 'version' command */
```
Comment (luisan00) [inline lib/cmd_system/src/cmd_system.c]: This isn't a valid documentation header

File: lib/cmd_system/src/cmd_system.c [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+static void register_free();
+static void register_heap();
+static void register_version();
+static void register_restart();
+static void register_deep_sleep();
+static void register_light_sleep();
+#if WITH_TASKS_INFO
+static void register_tasks();
+#endif
+
```
Comment (luisan00) [inline lib/cmd_system/src/cmd_system.c]: The documentation header is missing

File: lib/cmd_system/src/cmd_system.c [hunk 17]
```diff
@@ -0,0 +1,343 @@
+/* Console example — various system commands
+
+   This example code is in the Public Domain (or CC0 licensed, at your option.)
+
+   Unless required by applicable law or agreed to in writing, this
+   software is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+   CONDITIONS OF ANY KIND, either express or implied.
+*/
+
+#include <stdio.h>
```
Comment (luisan00) [inline lib/cmd_system/src/cmd_system.c]: When possible use `<cstdio>`

File: lib/cmd_system/src/cmd_system.c [hunk 18]
```diff
@@ -0,0 +1,343 @@
...
+#include "esp_sleep.h"
+#include "esp_spi_flash.h"
+#include "driver/rtc_io.h"
+#include "driver/uart.h"
+#include "argtable3/argtable3.h"
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "soc/rtc_cntl_reg.h"
+#include "esp32/rom/uart.h"
+
```
Comment (luisan00) [inline lib/cmd_system/src/cmd_system.c]: delete extra spacing

File: lib/cmd_system/src/cmd_system.c [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+}
+
+/** 'restart' command restarts the program */
+
+static int restart(int argc, char **argv)
+{
+    ESP_LOGI(TAG, "Restarting");
+    esp_restart();
+}
+
```
Comment (luisan00) [inline lib/cmd_system/src/cmd_system.c]: The documentation header is missing

File: lib/cmd_system/src/cmd_system.c [hunk 20]
```diff
@@ -0,0 +1,343 @@
...
+        .command = "restart",
+        .help = "Software reset of the chip",
+        .hint = NULL,
+        .func = &restart,
+    };
+    ESP_ERROR_CHECK( esp_console_cmd_register(&cmd) );
+}
+
+/** 'free' command prints available heap memory */
+
```
Comment (luisan00) [inline lib/cmd_system/src/cmd_system.c]: delete extra spacing

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 21]
```diff
@@ -0,0 +1,518 @@
...
+    for (int i = 0; i < TYPE_STR_PAIR_SIZE; i++) {
+        const type_str_pair_t *p = &type_str_pair[i];
+        if (strcmp(type, p->str) == 0) {
+            return  p->type;
+        }
+    }
+
+    return NVS_TYPE_ANY;
+}
+
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: The documentation header is missing

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 22]
```diff
@@ -0,0 +1,518 @@
...
+    if (err != ESP_OK) {
+        ESP_LOGE(TAG, "%s", esp_err_to_name(err));
+        return 1;
+    }
+
+    return 0;
+
+}
+
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: The documentation header is missing

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 23]
```diff
@@ -0,0 +1,518 @@
...
+    esp_err_t err = get_value_from_nvs(key, type);
+
+    if (err != ESP_OK) {
+        ESP_LOGE(TAG, "%s", esp_err_to_name(err));
+        return 1;
+    }
+
+    return 0;
+}
+
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: The documentation header is missing

File: test/test_serial/test_serial.c [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+
+#include <stdio.h>
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "driver/uart.h"
+#include "driver/gpio.h"
+#include <esp_log.h>
+
+
+/**
```
Comment (luisan00) [inline test/test_serial/test_serial.c]: doc. header in nowhere

File: test/test_serial/test_serial.c [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+ * it try to show how to work in a proper way with uart ports
+ * platformio test -e wemosbat -f test_serial
+ *  
+ */
+
+#define ECHO_TEST_TXD  (GPIO_NUM_21)
+#define ECHO_TEST_RXD  (GPIO_NUM_22)
+#define ECHO_TEST_RTS  (UART_PIN_NO_CHANGE)
+#define ECHO_TEST_CTS  (UART_PIN_NO_CHANGE)
+
```
Comment (luisan00) [inline test/test_serial/test_serial.c]: delete extra space

File: test/test_serial/test_serial.c [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+ * @license Apache 2.0, see LICENSE file for details
+ */
+
+#include <stdio.h>
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "driver/uart.h"
+#include "driver/gpio.h"
+#include <esp_log.h>
+
```
Comment (luisan00) [inline test/test_serial/test_serial.c]: delete extra space

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 27]
```diff
@@ -0,0 +1,518 @@
...
+static char current_namespace[16] = "storage";
+static const char *TAG = "cmd_nvs";
+
+static struct {
+    struct arg_str *key;
+    struct arg_str *type;
+    struct arg_str *value;
+    struct arg_end *end;
+} set_args;
+
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: doc.header is missing

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 28]
```diff
@@ -0,0 +1,518 @@
...
+static struct {
+    struct arg_str *key;
+    struct arg_str *type;
+    struct arg_end *end;
+} get_args;
+
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: doc.header is missing

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 29]
```diff
@@ -0,0 +1,518 @@
...
+static struct {
+    struct arg_str *key;
+    struct arg_end *end;
+} erase_args;
+
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: doc.header is missing

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 30]
```diff
@@ -0,0 +1,518 @@
...
+static const char *ARG_TYPE_STR = "type can be: i8, u8, i16, u16 i32, u32 i64, u64, str, blob";
+static char current_namespace[16] = "storage";
+static const char *TAG = "cmd_nvs";
+
+static struct {
+    struct arg_str *key;
+    struct arg_str *type;
+    struct arg_str *value;
+    struct arg_end *end;
+} set_args;
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: rename struct name `set_args` to `set_args_t`

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 31]
```diff
@@ -0,0 +1,518 @@
...
+
+static struct {
+    struct arg_str *key;
+    struct arg_str *type;
+    struct arg_end *end;
+} get_args;
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: rename to `get_args _t`

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 32]
```diff
@@ -0,0 +1,518 @@
...
+
+static struct {
+    struct arg_str *key;
+    struct arg_end *end;
+} erase_args;
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: rename to `erase_args_t`

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 33]
```diff
@@ -0,0 +1,518 @@
...
+
+static struct {
+    struct arg_str *namespace;
+    struct arg_end *end;
+} erase_all_args;
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: rename to `erase_all_args_t `

File: lib/cmd_nvs/src/cmd_nvs.c [hunk 34]
```diff
@@ -0,0 +1,518 @@
...
+
+static struct {
+    struct arg_str *namespace;
+    struct arg_end *end;
+} namespace_args;
```
Comment (luisan00) [inline lib/cmd_nvs/src/cmd_nvs.c]: rename to `namespace_args_t`

File: lib/cmd_system/src/cmd_system.c [hunk 35]
```diff
@@ -0,0 +1,343 @@
...
+   This example code is in the Public Domain (or CC0 licensed, at your option.)
+
+   Unless required by applicable law or agreed to in writing, this
+   software is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+   CONDITIONS OF ANY KIND, either express or implied.
+*/
+
+#include <stdio.h>
+#include <string.h>
```
Comment (luisan00) [inline lib/cmd_system/src/cmd_system.c]: use `<cstring>` instead

File: lib/cmd_system/src/cmd_system.c [hunk 36]
```diff
@@ -0,0 +1,343 @@
...
+    const esp_console_cmd_t cmd = {
+        .command = "restart",
+        .help = "Software reset of the chip",
+        .hint = NULL,
+        .func = &restart,
+    };
+    ESP_ERROR_CHECK( esp_console_cmd_register(&cmd) );
+}
+
+/** 'free' command prints available heap memory */
```
Comment (luisan00) [inline lib/cmd_system/src/cmd_system.c]: invalid inline comment

File: lib/cmd_system/src/cmd_system.c [hunk 37]
```diff
@@ -0,0 +1,343 @@
...
+
+static int free_mem(int argc, char **argv)
+{
+    printf("%d\n", esp_get_free_heap_size());
+    return 0;
+}
+
```
Comment (luisan00) [inline lib/cmd_system/src/cmd_system.c]: doc. header is missing

File: lib/cmd_system/src/cmd_system.c [hunk 38]
```diff
@@ -0,0 +1,343 @@
...
+#endif // WITH_TASKS_INFO
+
+/** 'deep_sleep' command puts the chip into deep sleep mode */
+
+static struct {
+    struct arg_int *wakeup_time;
+    struct arg_int *wakeup_gpio_num;
+    struct arg_int *wakeup_gpio_level;
+    struct arg_end *end;
+} deep_sleep_args;
```
Comment (luisan00) [inline lib/cmd_system/src/cmd_system.c]: use `deep_sleep_args_t` instead

**Comment by luisan00** [comment]
- Includes many static methods or code already available
- The cmd test is **only for testing purposes**  and it's not necessary to compile as a project library

**Comment by jeandudey** [comment]
@gustavosinbandera1 I don't know the purpose of this code, what is this for?

**Pull request closed without merging**